### PR TITLE
feat: add finalize_exit #15

### LINKED ISSUE #15: Write deposit/exit contract

### PR DESCRIPTION
**Pull request by syuhei176**
Close #15

### CODE REVIEW

File: contracts/actions/finalize_exit.ligo [hunk 1]
```diff
@@ -1,13 +1,87 @@
...
+    new_exit.subrange,
+    finalize_exit_params.deposited_range_id
+  );
+  // transfer
+  const state_object: property = decode_property(get_force(3n, new_exit.state_update.inputs));
+  const owner_opt: option(address) = decode_address(get_force(0n, state_object.inputs));
+  const withdraw_amount: int = new_exit.subrange.end_ - new_exit.subrange.start_;
+  const operations: list(operation) = list end;
+  case owner_opt of
+    | Some(owner) -> operations := transfer(owner, 1mutez(*abs(withdraw_amount)*)) # operations
```
Comment (syuhei176) [inline contracts/actions/finalize_exit.ligo]: Can I cast nat to tez?

Comment (syuhei176) [reply contracts/actions/finalize_exit.ligo]: solved

File: contracts/actions/finalize_exit.ligo [hunk 2]
```diff
@@ -1,13 +1,88 @@
...
+  const new_deposited_ranges: map(nat, range) = deposit_storage.deposited_ranges;
+  // check start
+  if range_to_remove.start_ =/= encompasing_range.start_
+  then deposit_storage.deposited_ranges[range_to_remove.end_] := record
+    start_ = encompasing_range.start_;
+    end_ = range_to_remove.end_;
+  end
+  else skip;
+  // check end
+  if range_to_remove.end_ = encompasing_range.start_
```
Comment (daiki44) [inline contracts/actions/finalize_exit.ligo]: `encompasing_range.end_`?

Comment (syuhei176) [reply contracts/actions/finalize_exit.ligo]: fixed and added test for this if condition.
https://github.com/cryptoeconomicslab/ovm.tz/pull/61/files#diff-4c8e9b9b5e7904409744e29b0b67c70cR205

File: contracts/actions/finalize_exit.ligo [hunk 3]
```diff
@@ -1,13 +1,90 @@
...
+  const exit_finalized_event: event_params = ExitFinalizedEvent(
+    (
+      finalize_exit_params.token_type,
+      get_exit_id(new_exit)
+    )
+  );
+  s.events_storage := emit_event(s.events_storage, "ExitFinalized", exit_finalized_event);
+end with ((operations : ops), s)
+
+function test_remove_deposited_range(
```
Comment (shogochiai) [inline contracts/actions/finalize_exit.ligo]: What will matter if we test with main entrypoint with Variant-typed arg?

Comment (syuhei176) [reply contracts/actions/finalize_exit.ligo]: Sorry, I don't know what you mean.

Comment (shogochiai) [reply contracts/actions/finalize_exit.ligo]: Pros
- Simpler initial_storage
- Higer resolution of the unit test

Cons
- Bigger contract size
- Duplication of initial_storage and it causes non-DRY-mess when a data structure modified.

Comment (syuhei176) [reply contracts/actions/finalize_exit.ligo]: you mean we don't need test for private method but public method?

Comment (syuhei176) [reply contracts/actions/finalize_exit.ligo]: as you write, I can't write enough test for public method now. because storage isn't separated.

Comment (syuhei176) [reply contracts/actions/finalize_exit.ligo]: I'll split entry point to fix "Bigger contract size"

Comment (shogochiai) [reply contracts/actions/finalize_exit.ligo]: I agree to go with your code.

FYI, IMHO, in the future if initial_storage is messing, we can abstract them via template DSL like https://github.com/MichelBoucey/glabrous

Comment (syuhei176) [reply contracts/actions/finalize_exit.ligo]: > I'll split entry point to fix "Bigger contract size"

added test entry point(test.ligo) to reduce main contract size.

**Pull request merged**